src/Microsoft.OData.Core/Evaluation/ODataConventionalResourceMetadataBuilder.cs

```diff
@@ -365,6 +368,26 @@ internal override void MarkNestedResourceInfoProcessed(string navigationProperty
             this.ProcessedNestedResourceInfos.Add(navigationPropertyName);
         }
 
+        /// <summary>
+        /// Marks the given expanded entity as processed.
+        /// </summary>
+        /// <param name="name">The name of the expanded entity we've already processed.</param>
+        internal override void MarkExpandedEntityProcessed(string expandedEntityName)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(expandedEntityName), "!string.IsNullOrEmpty(expandedEntityName)");
+            Debug.Assert(this.ProcessedExpandedEntityNames != null, "this.ProcessedExpandedEntityNames != null");
+            this.ProcessedExpandedEntityNames.Add(expandedEntityName);
+        }
+
+        /// <summary>
+        /// Return whether the given expanded entity of specified name has been processed.
+        /// </summary>
+        /// <param name="name">The name of the expanded entity.</param>
+        internal override bool IsExpandedEntityProcessed(string name)
+        {
+            return this.ProcessedExpandedEntityNames.Contains(name);
+        }
+

```

Comment:
Why not move this implementation to the ODataResourceMetadataBuilder as a (more useful) virtual implementation than a no-op for MarkExpandedEntityProcessed and returning false for IsExpandedEntityProcessed. A derived type can still override if they have alternate behavior.

Reply:
sure, moved.

---
In reply to: [229530080](https://github.com/OData/odata.net/pull/1106#discussion_r229530080) [](ancestors = 229530080)

Reply:
The implementation in ODataResourceMetadataBuilder just always returns false, and MarkExpandedEntityProcessed is a no-op. Why not put this actual implementation as the virtual implementation in ODataResourceMetadataBuilder, and not override at all in ODataConventionalResourceMetadataBuilder?

---
In reply to: [231309971](https://github.com/OData/odata.net/pull/1106#discussion_r231309971) [](ancestors = 231309971,229530080)